model/scope_test.go

```diff
@@ -0,0 +1,432 @@
+	for _, tc := range []struct {
+		name          string
+		have          []Scope
+		pluginID      string
+		expectedError string
+	}{
+		{
+			name:     "legacy app has access to all but internal",
+			have:     nil,
+			pluginID: "com.example.plugin",
+		},
+		{
+			name:     "empty have is same as legacy",
+			have:     []Scope{},
+			pluginID: "com.example.plugin",
+		},
+		{
+			name:     "exact match",
+			have:     []Scope{"channels:join", "plugins:com.example.plugin/somepath"},
+			pluginID: "com.example.plugin",
+		},
+		{
+			name:     "wildcard plugin id",
+			have:     []Scope{"channels:join", "plugins:*"},
+			pluginID: "com.example.plugin",
+		},
+		{
+			name:     "subpath",
+			have:     []Scope{"channels:join", "plugins:com.example.plugin/some"},
+			pluginID: "com.example.plugin",
+		},
+		{
+			name:     "no path prefix",
+			have:     []Scope{"channels:join", "plugins:com.example.plugin"},
+			pluginID: "com.example.plugin",
+		},
+		{
+			name:          "plugin ID mismatch",
+			have:          []Scope{"channels:join", "plugins:com.example.OTHER"},
+			pluginID:      "com.example.plugin",
+			expectedError: "insufficient scope, need plugins:com.example.plugin/somepath",
+		},
+		{
+			name:          "path mismatch",
+			have:          []Scope{"channels:join", "plugins:com.example.plugin/otherpath"},
+			pluginID:      "com.example.plugin",
+			expectedError: "insufficient scope, need plugins:com.example.plugin/somepath",
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			have := AppScopes(NormalizeScopes(tc.have))
+			err := have.SatisfiesPluginRequest(tc.pluginID, "/somepath")
+			if tc.expectedError != "" {
+				require.EqualError(t, err, tc.expectedError)
+			} else {
+				require.NoError(t, err)
+			}
+		})
+	}
+}
+func TestScopesSatisfies(t *testing.T) {
+	for _, tc := range []struct {
+		name       string
+		have, need []Scope
+		expected   bool
+	}{
+		{
+			name:     "legacy app has access to all but internal",
+			have:     nil,
+			need:     []Scope{"channels:join", "users:read"},
+			expected: true,
+		},
+		{
+			name:     "empty have is same as legacy",
+			have:     []Scope{},
+			need:     []Scope{"channels:join", "users:read"},
+			expected: true,
+		},
+		{
+			name:     "exact match",
+			have:     []Scope{"channels:join", "users:read"},
+			need:     []Scope{"channels:join", "users:read"},
+			expected: true,
+		},
+		{
+			name:     "superset",
+			have:     []Scope{"posts:read", "channels:join", "users:read"},
+			need:     []Scope{"channels:join", "users:read"},
+			expected: true,
+		},
+		{
+			name:     "have has a global wildcard",
+			have:     []Scope{"*:*", "channels:join", "users:read"},
+			need:     []Scope{"channels:join", "users:read"},
+			expected: true,
+		},
+		{
+			name:     "need has a global wildcard",
+			have:     []Scope{"channels:join", "users:read"},
+			need:     []Scope{"*:*", "channels:join", "users:read"},
+			expected: true,
```

Comment:
Should this fail instead? It seems `have` does not satisfy `need` but I may be misundertanding

Reply:
A `*:*` in the `APIScopes` means that the API is "unrestricted". All other scopes will get normalized out when it is parsed.